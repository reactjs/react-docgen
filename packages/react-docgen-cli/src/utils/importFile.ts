import { createRequire } from 'module';
import { pathToFileURL } from 'url';

const require = createRequire(import.meta.url);
const resolveOptions = { paths: [process.cwd()] };

export default async function importFile<T>(
  importSpecifier: string,
): Promise<T | undefined> {
  try {
    const importedFile = await import(
      // need to convert to file:// url as on windows absolute path strings do not work
      pathToFileURL(require.resolve(importSpecifier, resolveOptions)).href
    );

    return importedFile.default ? importedFile.default : importedFile;
  } catch (error) {
    if (
      error instanceof Error &&
      (error as NodeJS.ErrnoException).code !== 'ERR_MODULE_NOT_FOUND' &&
      (error as NodeJS.ErrnoException).code !== 'MODULE_NOT_FOUND'
    ) {
      throw error;
    }

    return undefined;
  }
}
