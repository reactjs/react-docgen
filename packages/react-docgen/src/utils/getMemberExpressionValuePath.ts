import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import type { Expression } from '@babel/types';
import getNameOrValue from './getNameOrValue.js';
import { String as toString } from './expressionTo.js';
import isReactForwardRefCall from './isReactForwardRefCall.js';

function resolveName(path: NodePath): string | undefined {
  if (path.isVariableDeclaration()) {
    const declarations = path.get('declarations');

    if (declarations.length > 1) {
      throw new TypeError(
        'Got unsupported VariableDeclaration. VariableDeclaration must only ' +
          'have a single VariableDeclarator. Got ' +
          declarations.length +
          ' declarations.',
      );
    }
    // VariableDeclarator always has at least one declaration, hence the non-null-assertion
    const id = declarations[0]!.get('id');

    if (id.isIdentifier()) {
      return id.node.name;
    }

    return;
  }

  if (path.isFunctionDeclaration()) {
    const id = path.get('id');

    if (id.isIdentifier()) {
      return id.node.name;
    }

    return;
  }

  // When we check ObjectMethod we simply ignore it as assigning
  // to it is technicaly possible but rare
  if (path.isObjectMethod()) {
    return;
  }

  if (
    path.isFunctionExpression() ||
    path.isArrowFunctionExpression() ||
    path.isTaggedTemplateExpression() ||
    path.isCallExpression() ||
    isReactForwardRefCall(path)
  ) {
    let currentPath: NodePath = path;

    while (currentPath.parentPath) {
      if (currentPath.parentPath.isVariableDeclarator()) {
        const id = currentPath.parentPath.get('id');

        if (id.isIdentifier()) {
          return id.node.name;
        }

        return;
      }

      currentPath = currentPath.parentPath;
    }

    return;
  }

  throw new TypeError(
    'Attempted to resolveName for an unsupported path. resolveName does not accept ' +
      path.node.type +
      '".',
  );
}

interface TraverseState {
  readonly memberName: string;
  readonly localName: string;
  result: NodePath<Expression> | null;
}

const explodedVisitors = visitors.explode<TraverseState>({
  AssignmentExpression: {
    enter: function (path, state) {
      const memberPath = path.get('left');

      if (!memberPath.isMemberExpression()) {
        return;
      }
      const property = memberPath.get('property');

      if (
        ((!memberPath.node.computed && property.isIdentifier()) ||
          property.isStringLiteral() ||
          property.isNumericLiteral()) &&
        getNameOrValue(property) === state.memberName &&
        toString(memberPath.get('object')) === state.localName
      ) {
        state.result = path.get('right');
        path.stop();
      }
    },
  },
});

export default function getMemberExpressionValuePath(
  variableDefinition: NodePath,
  memberName: string,
): NodePath<Expression> | null {
  const localName = resolveName(variableDefinition);

  if (!localName) {
    // likely an immediately exported and therefore nameless/anonymous node
    // passed in
    return null;
  }

  const state: TraverseState = {
    localName,
    memberName,
    result: null,
  };

  variableDefinition.hub.file.traverse(explodedVisitors, state);

  return state.result;
}
