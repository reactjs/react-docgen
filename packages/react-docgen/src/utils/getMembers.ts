import type { NodePath } from '@babel/traverse';
import type { Expression, PrivateName } from '@babel/types';

interface MemberDescriptor {
  path: NodePath<Expression | PrivateName>;
  computed: boolean;
  argumentPaths: NodePath[];
}

/**
 * Given a "nested" Member/CallExpression, e.g.
 *
 * foo.bar()[baz][42]
 *
 * this returns a list of "members". In this example it would be something like
 * [
 *   {path: NodePath<bar>, arguments: NodePath<empty>, computed: false},
 *   {path: NodePath<baz>, arguments: null, computed: true},
 *   {path: NodePath<42>, arguments: null, computed: false}
 * ]
 */
export default function getMembers(
  path: NodePath,
  includeRoot = false,
): MemberDescriptor[] {
  const result: MemberDescriptor[] = [];
  let argumentPaths: NodePath[] = [];
  let resultPath: MemberDescriptor['path'] = path as NodePath<Expression>;

  while (true) {
    if (resultPath.isMemberExpression()) {
      const property = resultPath.get('property');

      result.push({
        path: property,
        computed: resultPath.node.computed,
        argumentPaths,
      });
      argumentPaths = [];
      resultPath = resultPath.get('object');
    } else if (resultPath.isCallExpression()) {
      const callee = resultPath.get('callee');

      if (callee.isExpression()) {
        argumentPaths = resultPath.get('arguments');
        resultPath = callee;
      } else {
        break;
      }
    } else {
      break;
    }
  }
  if (includeRoot && result.length > 0) {
    result.push({
      path: resultPath,
      computed: false,
      argumentPaths,
    });
  }

  return result.reverse();
}
