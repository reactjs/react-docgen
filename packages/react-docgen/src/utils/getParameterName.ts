import type { NodePath } from '@babel/traverse';
import type { FunctionParameter, TSParameterProperty } from '@babel/types';
import printValue from './printValue.js';

type ParameterNodePath = NodePath<FunctionParameter | TSParameterProperty>;

export default function getParameterName(
  parameterPath: ParameterNodePath,
): string {
  if (parameterPath.isIdentifier()) {
    return parameterPath.node.name;
  } else if (parameterPath.isAssignmentPattern()) {
    return getParameterName(parameterPath.get('left') as ParameterNodePath);
  } else if (
    parameterPath.isObjectPattern() ||
    parameterPath.isArrayPattern()
  ) {
    return printValue(parameterPath);
  } else if (parameterPath.isRestElement()) {
    return `...${getParameterName(
      parameterPath.get('argument') as ParameterNodePath,
    )}`;
  } else if (parameterPath.isTSParameterProperty()) {
    return getParameterName(parameterPath.get('parameter'));
  } else if (parameterPath.isVoidPattern()) {
    return 'void';
  }

  throw new TypeError(
    'Parameter name must be one of Identifier, AssignmentPattern, ArrayPattern, ' +
      `ObjectPattern, RestElement, or VoidPattern instead got ${parameterPath.node.type}`,
  );
}
