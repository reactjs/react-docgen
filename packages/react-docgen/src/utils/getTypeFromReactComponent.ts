import type { NodePath } from '@babel/traverse';
import type Documentation from '../Documentation.js';
import getMemberValuePath from './getMemberValuePath.js';
import getTypeAnnotation from './getTypeAnnotation.js';
import getTypeParameters from './getTypeParameters.js';
import isReactComponentClass from './isReactComponentClass.js';
import isReactForwardRefCall from './isReactForwardRefCall.js';
import resolveGenericTypeAnnotation from './resolveGenericTypeAnnotation.js';
import resolveToValue from './resolveToValue.js';
import type { TypeParameters } from './getTypeParameters.js';
import type {
  FlowType,
  InterfaceDeclaration,
  InterfaceExtends,
  TSExpressionWithTypeArguments,
  TSInterfaceDeclaration,
  TSType,
  TSTypeParameterDeclaration,
  TSTypeParameterInstantiation,
  TypeParameterDeclaration,
  TypeParameterInstantiation,
  VariableDeclarator,
} from '@babel/types';
import getTypeIdentifier from './getTypeIdentifier.js';
import isReactBuiltinReference from './isReactBuiltinReference.js';
import unwrapBuiltinTSPropTypes from './unwrapBuiltinTSPropTypes.js';

function getStatelessPropsPath(
  componentDefinition: NodePath,
): NodePath | undefined {
  if (!componentDefinition.isFunction()) return;

  return componentDefinition.get('params')[0];
}

function getForwardRefGenericsType(
  componentDefinition: NodePath,
): NodePath<TSType> | null {
  const typeParameters = componentDefinition.get('typeParameters') as NodePath<
    TSTypeParameterInstantiation | null | undefined
  >;

  if (typeParameters && typeParameters.hasNode()) {
    const params = typeParameters.get('params');

    return params[1] ?? null;
  }

  return null;
}

function findAssignedVariableType(
  componentDefinition: NodePath,
): NodePath | null {
  const variableDeclarator = componentDefinition.findParent((path) =>
    path.isVariableDeclarator(),
  ) as NodePath<VariableDeclarator> | null;

  if (!variableDeclarator) return null;

  const typeAnnotation = getTypeAnnotation(variableDeclarator.get('id'));

  if (!typeAnnotation) return null;

  if (typeAnnotation.isTSTypeReference()) {
    const typeName = typeAnnotation.get('typeName');

    if (
      isReactBuiltinReference(typeName, 'FunctionComponent') ||
      isReactBuiltinReference(typeName, 'FC') ||
      isReactBuiltinReference(typeName, 'VoidFunctionComponent') ||
      isReactBuiltinReference(typeName, 'VFC')
    ) {
      const typeParameters = typeAnnotation.get('typeParameters');

      if (typeParameters.hasNode()) {
        return typeParameters.get('params')[0] ?? null;
      }
    }
  }

  return null;
}

/**
 * Given an React component (stateless or class) tries to find
 * flow or TS types for the props. It may find multiple types.
 * If not found or it is not one of the supported component types,
 *  this function returns an empty array.
 */
export default (componentDefinition: NodePath): NodePath[] => {
  const typePaths: NodePath[] = [];

  if (isReactComponentClass(componentDefinition)) {
    const superTypes = componentDefinition.get('superTypeParameters');

    if (superTypes.hasNode()) {
      const params = superTypes.get('params');

      if (params.length >= 1) {
        typePaths.push(params[params.length === 3 ? 1 : 0]!);
      }
    } else {
      const propsMemberPath = getMemberValuePath(componentDefinition, 'props');

      if (!propsMemberPath) {
        return [];
      }

      const typeAnnotation = getTypeAnnotation(propsMemberPath.parentPath);

      if (typeAnnotation) {
        typePaths.push(typeAnnotation);
      }
    }
  } else {
    if (isReactForwardRefCall(componentDefinition)) {
      const genericTypeAnnotation =
        getForwardRefGenericsType(componentDefinition);

      if (genericTypeAnnotation) {
        typePaths.push(genericTypeAnnotation);
      }

      componentDefinition = resolveToValue(
        componentDefinition.get('arguments')[0]!,
      );
    }

    const propsParam = getStatelessPropsPath(componentDefinition);

    if (propsParam) {
      const typeAnnotation = getTypeAnnotation(propsParam);

      if (typeAnnotation) {
        typePaths.push(typeAnnotation);
      }
    }

    const assignedVariableType = findAssignedVariableType(componentDefinition);

    if (assignedVariableType) {
      typePaths.push(assignedVariableType);
    }
  }

  return typePaths.map((typePath) => unwrapBuiltinTSPropTypes(typePath));
};

export function applyToTypeProperties(
  documentation: Documentation,
  path: NodePath,
  callback: (propertyPath: NodePath, params: TypeParameters | null) => void,
  typeParams: TypeParameters | null,
): void {
  if (path.isObjectTypeAnnotation()) {
    path
      .get('properties')
      .forEach((propertyPath) => callback(propertyPath, typeParams));
  } else if (path.isTSTypeLiteral()) {
    path
      .get('members')
      .forEach((propertyPath) => callback(propertyPath, typeParams));
  } else if (path.isInterfaceDeclaration()) {
    applyExtends(documentation, path, callback, typeParams);

    path
      .get('body')
      .get('properties')
      .forEach((propertyPath) => callback(propertyPath, typeParams));
  } else if (path.isTSInterfaceDeclaration()) {
    applyExtends(documentation, path, callback, typeParams);

    path
      .get('body')
      .get('body')
      .forEach((propertyPath) => callback(propertyPath, typeParams));
  } else if (
    path.isIntersectionTypeAnnotation() ||
    path.isTSIntersectionType()
  ) {
    (path.get('types') as Array<NodePath<FlowType | TSType>>).forEach(
      (typesPath) =>
        applyToTypeProperties(documentation, typesPath, callback, typeParams),
    );
  } else if (!path.isUnionTypeAnnotation()) {
    // The react-docgen output format does not currently allow
    // for the expression of union types
    const typePath = resolveGenericTypeAnnotation(path);

    if (typePath) {
      applyToTypeProperties(documentation, typePath, callback, typeParams);
    }
  }
}

function applyExtends(
  documentation: Documentation,
  path: NodePath<InterfaceDeclaration | TSInterfaceDeclaration>,
  callback: (propertyPath: NodePath, params: TypeParameters | null) => void,
  typeParams: TypeParameters | null,
): void {
  const classExtends = path.get('extends');

  if (!Array.isArray(classExtends)) {
    return;
  }

  classExtends.forEach(
    (
      extendsPath: NodePath<InterfaceExtends | TSExpressionWithTypeArguments>,
    ) => {
      const resolvedPath = resolveGenericTypeAnnotation(extendsPath);

      if (resolvedPath) {
        if (
          resolvedPath.has('typeParameters') &&
          extendsPath.node.typeParameters
        ) {
          typeParams = getTypeParameters(
            resolvedPath.get('typeParameters') as NodePath<
              TSTypeParameterDeclaration | TypeParameterDeclaration
            >,
            extendsPath.get('typeParameters') as NodePath<
              TSTypeParameterInstantiation | TypeParameterInstantiation
            >,
            typeParams,
          );
        }
        applyToTypeProperties(
          documentation,
          resolvedPath,
          callback,
          typeParams,
        );
      } else {
        const idPath = getTypeIdentifier(extendsPath);

        if (idPath && idPath.isIdentifier()) {
          documentation.addComposes(idPath.node.name);
        }
      }
    },
  );
}
