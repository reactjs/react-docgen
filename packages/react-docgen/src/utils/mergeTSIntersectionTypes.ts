import type {
  TSFunctionSignatureType,
  TypeDescriptor,
} from '../Documentation.js';

/**
 * Merges two TSFunctionSignatureType types into one.
 *
 * @example
 * const existingType = {
 *   "key": "children",
 *   "value": {
 *     "name": "ReactNode",
 *     "required": true,
 *   },
 * };
 * const newType = {
 *   "key": "children",
 *   "value": {
 *     "name": "never",
 *     "required": false,
 *   },
 * };
 *
 * return {
 *   "key": "children",
 *   "value": {
 *     "name": "ReactNode",
 *     "required": false,
 *   },
 * };
 */
export default (
  existingType: TypeDescriptor<TSFunctionSignatureType>,
  newType: TypeDescriptor<TSFunctionSignatureType>,
): TypeDescriptor<TSFunctionSignatureType> => {
  const required =
    newType.required === false || existingType.required === false
      ? false
      : existingType.required;

  const existingTypesArray = existingType.name.split('|').map((t) => t.trim());
  const existingTypes = new Set(existingTypesArray);

  if (!['never'].includes(newType.name)) {
    existingTypes.add(newType.name);
  }

  if (existingType.name === 'unknown' || newType.name === 'unknown') {
    return {
      name: 'unknown',
      required,
    };
  }

  return {
    name: Array.from(existingTypes).join(' | '),
    required,
  };
};
