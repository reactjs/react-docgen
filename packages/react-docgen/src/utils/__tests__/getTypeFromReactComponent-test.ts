import type {
  ArrowFunctionExpression,
  ClassDeclaration,
  VariableDeclaration,
} from '@babel/types';
import { parse, parseTypescript } from '../../../tests/utils';
import getTypeFromReactComponent from '../getTypeFromReactComponent.js';
import { describe, expect, test } from 'vitest';
import type { NodePath } from '@babel/traverse';

describe('getTypeFromReactComponent', () => {
  test('handles no stateless props', () => {
    const path = parseTypescript
      .statementLast<VariableDeclaration>(`const x = () => {}`)
      .get('declarations')[0]
      .get('init') as NodePath<ArrowFunctionExpression>;

    expect(getTypeFromReactComponent(path)).toMatchSnapshot();
  });

  test('handles no class props', () => {
    const path = parseTypescript.statementLast<ClassDeclaration>(
      `import React from 'react';
       class X extends React.Component {
         render() {}
       }`,
    );

    expect(getTypeFromReactComponent(path)).toMatchSnapshot();
  });

  describe('TypeScript', () => {
    describe('stateless', () => {
      test('finds param type annotation', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(`const x = (props: Props) => {}`)
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds wrapped param type annotation', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(
            `import React from 'react';
             const x = (props: React.PropsWithChildren<Props>) => {}`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds generic forwardRef type annotation', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(
            `import React from 'react';
             const x = React.forwardRef<HTMLDivElement, React.PropsWithChildren<Props>>((props, ref) => {})`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('does not find generic forwardRef type annotation on typo', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(
            `import React from 'react';
             const x = React.backwardRef<HTMLDivElement, React.PropsWithChildren<Props>>((props, ref) => {})`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds param inline type', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(
            `const x = (props: { prop: string }) => {}`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      describe.each([
        'FunctionComponent',
        'FC',
        'VoidFunctionComponent',
        'VFC',
      ])('finds variable type annotation (%s)', (name) => {
        test('with MemberExpression', () => {
          const path = parseTypescript
            .statementLast<VariableDeclaration>(
              `import React from 'react';
               const x: React.${name}<Props> = (props) => {}`,
            )
            .get('declarations')[0]
            .get('init') as NodePath<ArrowFunctionExpression>;

          expect(getTypeFromReactComponent(path)).toMatchSnapshot();
        });

        test('with named import', () => {
          const path = parseTypescript
            .statementLast<VariableDeclaration>(
              `import { ${name} } from 'react';
               const x: ${name}<Props> = (props) => {}`,
            )
            .get('declarations')[0]
            .get('init') as NodePath<ArrowFunctionExpression>;

          expect(getTypeFromReactComponent(path)).toMatchSnapshot();
        });
      });

      test('finds multiple variable type annotation', () => {
        const path = parseTypescript
          .statementLast<VariableDeclaration>(
            `import React from 'react';
             const x: React.FC<Props> = (props: Props) => {}`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });
    });

    describe('classes', () => {
      test('finds props type in params', () => {
        const path = parseTypescript.statementLast<ClassDeclaration>(
          `import React from 'react';
             class X extends React.Component<Props, State> {
               render() {}
             }`,
        );

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds props type in properties', () => {
        const path = parseTypescript.statementLast<ClassDeclaration>(
          `import React from 'react';
           class X extends React.Component {
             props: Props;
             render() {}
           }`,
        );

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });
    });
  });

  describe('Flow', () => {
    describe('stateless', () => {
      test('finds param type annotation', () => {
        const path = parse
          .statementLast<VariableDeclaration>(`const x = (props: Props) => {}`)
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds param inline type', () => {
        const path = parse
          .statementLast<VariableDeclaration>(
            `const x = (props: { prop: string }) => {}`,
          )
          .get('declarations')[0]
          .get('init') as NodePath<ArrowFunctionExpression>;

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });
    });

    describe('classes', () => {
      test('finds props type in new params', () => {
        const path = parse.statementLast<ClassDeclaration>(
          `import React from 'react';
             class X extends React.Component<Props, State> {
               render() {}
             }`,
        );

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds props type in old params', () => {
        const path = parse.statementLast<ClassDeclaration>(
          `import React from 'react';
             class X extends React.Component<DefaultProps, Props, State> {
               render() {}
             }`,
        );

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });

      test('finds props type in properties', () => {
        const path = parse.statementLast<ClassDeclaration>(
          `import React from 'react';
           class X extends React.Component {
             props: Props;
             render() {}
           }`,
        );

        expect(getTypeFromReactComponent(path)).toMatchSnapshot();
      });
    });
  });
});
