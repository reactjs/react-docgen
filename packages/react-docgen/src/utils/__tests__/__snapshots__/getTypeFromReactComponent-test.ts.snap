// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTypeFromReactComponent > Flow > classes > finds props type in new params 1`] = `
[
  Node {
    "id": Node {
      "name": "Props",
      "type": "Identifier",
    },
    "type": "GenericTypeAnnotation",
    "typeParameters": null,
  },
]
`;

exports[`getTypeFromReactComponent > Flow > classes > finds props type in old params 1`] = `
[
  Node {
    "id": Node {
      "name": "Props",
      "type": "Identifier",
    },
    "type": "GenericTypeAnnotation",
    "typeParameters": null,
  },
]
`;

exports[`getTypeFromReactComponent > Flow > classes > finds props type in properties 1`] = `
[
  Node {
    "id": Node {
      "name": "Props",
      "type": "Identifier",
    },
    "type": "GenericTypeAnnotation",
    "typeParameters": null,
  },
]
`;

exports[`getTypeFromReactComponent > Flow > stateless > finds param inline type 1`] = `
[
  Node {
    "callProperties": [],
    "exact": false,
    "indexers": [],
    "inexact": false,
    "internalSlots": [],
    "properties": [
      Node {
        "key": Node {
          "name": "prop",
          "type": "Identifier",
        },
        "kind": "init",
        "method": false,
        "optional": false,
        "proto": false,
        "static": false,
        "type": "ObjectTypeProperty",
        "value": Node {
          "type": "StringTypeAnnotation",
        },
        "variance": null,
      },
    ],
    "type": "ObjectTypeAnnotation",
  },
]
`;

exports[`getTypeFromReactComponent > Flow > stateless > finds param type annotation 1`] = `
[
  Node {
    "id": Node {
      "name": "Props",
      "type": "Identifier",
    },
    "type": "GenericTypeAnnotation",
    "typeParameters": null,
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > classes > finds props type in params 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > classes > finds props type in properties 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > does not find generic forwardRef type annotation on typo 1`] = `[]`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds generic forwardRef type annotation 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds multiple variable type annotation 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds param inline type 1`] = `
[
  Node {
    "members": [
      Node {
        "computed": false,
        "key": Node {
          "name": "prop",
          "type": "Identifier",
        },
        "type": "TSPropertySignature",
        "typeAnnotation": Node {
          "type": "TSTypeAnnotation",
          "typeAnnotation": Node {
            "type": "TSStringKeyword",
          },
        },
      },
    ],
    "type": "TSTypeLiteral",
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds param type annotation 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (FC) > with MemberExpression 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (FC) > with named import 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (FunctionComponent) > with MemberExpression 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (FunctionComponent) > with named import 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (VFC) > with MemberExpression 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (VFC) > with named import 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (VoidFunctionComponent) > with MemberExpression 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds variable type annotation (VoidFunctionComponent) > with named import 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > TypeScript > stateless > finds wrapped param type annotation 1`] = `
[
  Node {
    "type": "TSTypeReference",
    "typeName": Node {
      "name": "Props",
      "type": "Identifier",
    },
  },
]
`;

exports[`getTypeFromReactComponent > handles no class props 1`] = `[]`;

exports[`getTypeFromReactComponent > handles no stateless props 1`] = `[]`;
